mono/sgen/sgen-los.c

```diff
@@ -432,39 +487,35 @@ static void sgen_los_unpin_object (GCObject *data);
 void
 sgen_los_sweep (void)
 {
-	LOSObject *bigobj, *prevbo;
+	LOSObject *obj;
 	LOSSection *section, *prev;
 	int i;
 	int num_sections = 0;
 
 	/* sweep the big objects list */
-	prevbo = NULL;
-	for (bigobj = sgen_los_object_list; bigobj;) {
-		SGEN_ASSERT (0, !SGEN_OBJECT_IS_PINNED (bigobj->data), "Who pinned a LOS object?");
-
-		if (sgen_los_object_is_pinned (bigobj->data)) {
-			if (bigobj->cardtable_mod_union) {
-				mword obj_size = sgen_los_object_size (bigobj);
-				mword num_cards = sgen_card_table_number_of_cards_in_range ((mword) bigobj->data, obj_size);
-				memset (bigobj->cardtable_mod_union, 0, num_cards);
+	FOREACH_LOS_OBJECT_NO_LOCK (obj) {
+		SGEN_ASSERT (0, !SGEN_OBJECT_IS_PINNED (obj->data), "Who pinned a LOS object?");
```

Comment:
Can a user do this? (Users can't tell a large object apart from a small object, right?)  Or does Sgen just ignore requests to pin/unpin large objects?

Reply:
This pinning is internal to the GC and it is only set during a collection. When the mutator wants to pin an object it allocates a pinned gchandle, which means the reference to that object is stored in an sgen structure that is internally registered as a conservative root (see ROOT_TYPE_PINNED coupled with MONO_ROOT_SOURCE_GC_HANDLE). During the collection the GC will pin all objects encountered there and make sure they are unpinned when the collection is done. The objects will remain in the other structure until the user deregisters the gchandle.

Reply:
I don't think we ignore request to pin LOS objects, even though we technically could.

Reply:
Found this in sgen_alloc_obj_pinned: